pkg/server/server_startup_test.go

```diff
@@ -0,0 +1,310 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package server_test
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/server"
+	"github.com/cockroachdb/cockroach/pkg/spanconfig"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
+	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/randutil"
+	"github.com/cockroachdb/errors"
+	"github.com/stretchr/testify/require"
+)
+
+// TestStartupFailure can help reproducing failures found in randomized test as
+// it may be tricky to get the same results easily. To reproduce failure, use
+// failing range from test and plug in into faultKey.
+// As of time of writing following ranges trigger retries:
+// - /System/NodeLivenessMax (keys.BootstrapVersionKey)
+// - /Table/11 (keys.SystemSQLCodec.TablePrefix(11))
+// If startup.IsRetryableReplicaError is changed to return false every time,
+// test should fail if range from above is used.
+func TestStartupFailure(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	skip.IgnoreLint(t, "this test should be used to reproduce failures from random tests")
+
+	// Set faultyKey to suspected range that is causing startup failure, comment
+	// out skip statement above and run the test.
+	faultyKey := keys.BootstrapVersionKey
+	runCircuitBreakerTestForKey(t,
+		func(spans []roachpb.Span) (good, bad []roachpb.Key) {
+			for _, span := range spans {
+				if span.ContainsKey(faultyKey) {
+					bad = append(bad, span.Key)
+				} else {
+					good = append(good, span.Key)
+				}
+			}
+			return good, bad
+		})
+}
+
+// TestStartupFailureRandomRange picks one random range and runs the test on it.
+// If this test failed, then suspicious range start key deduced from error could
+// be used in TestStartupFailure to investigate code path not covered by startup
+// retries.
+func TestStartupFailureRandomRange(t *testing.T) {
```

Comment:
```
--- PASS: TestStartupFailureRandomRange (31.04s)
```

Spending 30+s on a single test seems difficult to justify, especially since we also have `TestStartupInjectedFailureSingleNode` which takes less than 2s and which should provide similar coverage.

I'm even hesitant to merge this for only nightly stress, since spending 30s on each iteration of the package on this test cannibalizes other tests and reduces our chances of finding flakes.

We could move this test into a leaf package (`startup_test`, i.e. in `startup`?) to avoid this problem - the `startup` package would basically be stressing these new tests only. Is there a dependency that would prohibit this (i.e. are you using anything in `server_test` in this test)? Note that this test should continue to run only under stress; we don't want a 30+s test on PR/bors CI.